README.rst
Python Client for Network Security API API
==========================================

|preview| |pypi| |versions|

`Network Security API API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |preview| image:: https://img.shields.io/badge/support-preview-orange.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-network-security.svg
   :target: https://pypi.org/project/google-cloud-network-security/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-network-security.svg
   :target: https://pypi.org/project/google-cloud-network-security/
.. _Network Security API API: https://cloud.google.com/traffic-director/docs/reference/network-security/rest
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/networksecurity/latest
.. _Product Documentation:  https://cloud.google.com/traffic-director/docs/reference/network-security/rest

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Network Security API API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Network Security API API.:  https://cloud.google.com/traffic-director/docs/reference/network-security/rest
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-network-security


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-network-security

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Network Security API API
   to see other available methods on the client.
-  Read the `Network Security API API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Network Security API API Product documentation:  https://cloud.google.com/traffic-director/docs/reference/network-security/rest
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [0.6.1](https://github.com/googleapis/python-network-security/compare/v0.6.0...v0.6.1) (2022-07-23)


### Bug Fixes

* fix annotation of parent in Create*Request ([f53ca2d](https://github.com/googleapis/python-network-security/commit/f53ca2d453468f81eb8adf96e58125e7629ca634))


### Documentation

* update the comments of various networksecurity resources ([#115](https://github.com/googleapis/python-network-security/issues/115)) ([f53ca2d](https://github.com/googleapis/python-network-security/commit/f53ca2d453468f81eb8adf96e58125e7629ca634))

## [0.6.0](https://github.com/googleapis/python-network-security/compare/v0.5.0...v0.6.0) (2022-07-16)


### Features

* add network_security_v1 ([c2694d8](https://github.com/googleapis/python-network-security/commit/c2694d8c07a92fed69a6b4363fc2c70ff4d14022))


### Bug Fixes

* update default import for network_security to network_security_v1 ([c2694d8](https://github.com/googleapis/python-network-security/commit/c2694d8c07a92fed69a6b4363fc2c70ff4d14022))

## [0.5.0](https://github.com/googleapis/python-network-security/compare/v0.4.2...v0.5.0) (2022-07-13)


### Features

* add audience parameter ([5e99e7e](https://github.com/googleapis/python-network-security/commit/5e99e7ed2ed3ac60009d237ab8131ed6ca0725a9))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#110](https://github.com/googleapis/python-network-security/issues/110)) ([a700dc2](https://github.com/googleapis/python-network-security/commit/a700dc2c70e91e92540d819e412a91b1253e2f40))
* require python 3.7+ ([#108](https://github.com/googleapis/python-network-security/issues/108)) ([dea6632](https://github.com/googleapis/python-network-security/commit/dea6632f8b376ec619260f2207d996f1994d46e5))

## [0.4.2](https://github.com/googleapis/python-network-security/compare/v0.4.1...v0.4.2) (2022-06-03)


### Bug Fixes

* **deps:** require protobuf <4.0.0dev ([#98](https://github.com/googleapis/python-network-security/issues/98)) ([dfdc5c7](https://github.com/googleapis/python-network-security/commit/dfdc5c706382ba14a2a12a8e28530abcd8b48724))


### Documentation

* fix changelog header to consistent size ([#99](https://github.com/googleapis/python-network-security/issues/99)) ([86bc9e4](https://github.com/googleapis/python-network-security/commit/86bc9e4db6a2332ace0a8f6d8a194261f38dd072))

## [0.4.1](https://github.com/googleapis/python-network-security/compare/v0.4.0...v0.4.1) (2022-03-05)


### Bug Fixes

* **deps:** require google-api-core>=1.31.5, >=2.3.2 ([#77](https://github.com/googleapis/python-network-security/issues/77)) ([6f0a9cb](https://github.com/googleapis/python-network-security/commit/6f0a9cbd206d99a8d4b1f3a3af5e8da1c88c838c))

## [0.4.0](https://github.com/googleapis/python-network-security/compare/v0.3.2...v0.4.0) (2022-02-26)


### Features

* add api key support ([#63](https://github.com/googleapis/python-network-security/issues/63)) ([3f15e76](https://github.com/googleapis/python-network-security/commit/3f15e7688638adfa55eda09c466cb75e7a793b12))


### Bug Fixes

* resolve DuplicateCredentialArgs error when using credentials_file ([a531e63](https://github.com/googleapis/python-network-security/commit/a531e6371651a5036afbc82d6ee10c1bc50e7610))

## [0.3.2](https://www.github.com/googleapis/python-network-security/compare/v0.3.1...v0.3.2) (2022-01-09)


### Bug Fixes

* updating metadata messages for all long running operations ([#57](https://www.github.com/googleapis/python-network-security/issues/57)) ([7adc601](https://www.github.com/googleapis/python-network-security/commit/7adc601de611fe0323185b2747c98a620e21a38f))

## [0.3.1](https://www.github.com/googleapis/python-network-security/compare/v0.3.0...v0.3.1) (2021-11-02)


### Bug Fixes

* **deps:** drop packaging dependency ([518f32b](https://www.github.com/googleapis/python-network-security/commit/518f32b90db80cd8a5b2774aba8c9a4b13ea1f57))
* **deps:** require google-api-core >= 1.28.0 ([518f32b](https://www.github.com/googleapis/python-network-security/commit/518f32b90db80cd8a5b2774aba8c9a4b13ea1f57))


### Documentation

* list oneofs in docstring ([518f32b](https://www.github.com/googleapis/python-network-security/commit/518f32b90db80cd8a5b2774aba8c9a4b13ea1f57))

## [0.3.0](https://www.github.com/googleapis/python-network-security/compare/v0.2.0...v0.3.0) (2021-10-14)


### Features

* add support for python 3.10 ([#41](https://www.github.com/googleapis/python-network-security/issues/41)) ([7bfb72f](https://www.github.com/googleapis/python-network-security/commit/7bfb72f3e58685ed588b14f855beb8630e0eabd5))

## [0.2.0](https://www.github.com/googleapis/python-network-security/compare/v0.1.5...v0.2.0) (2021-10-08)


### Features

* add context manager support in client ([#37](https://www.github.com/googleapis/python-network-security/issues/37)) ([173b3db](https://www.github.com/googleapis/python-network-security/commit/173b3dbd36c5118853b8d93dcb32635d64208876))

## [0.1.5](https://www.github.com/googleapis/python-network-security/compare/v0.1.4...v0.1.5) (2021-09-30)


### Bug Fixes

* improper types in pagers generation ([90b83d6](https://www.github.com/googleapis/python-network-security/commit/90b83d6282d8b68890eed8e81013766763ec6648))

## [0.1.4](https://www.github.com/googleapis/python-network-security/compare/v0.1.3...v0.1.4) (2021-09-24)


### Bug Fixes

* add 'dict' annotation type to 'request' ([dcefd7e](https://www.github.com/googleapis/python-network-security/commit/dcefd7e92f08e12c868fee114c73075e8ba356a5))

## [0.1.3](https://www.github.com/googleapis/python-network-security/compare/v0.1.2...v0.1.3) (2021-07-29)


### Bug Fixes

* enable self signed jwt for grpc ([#15](https://www.github.com/googleapis/python-network-security/issues/15)) ([1cea5e5](https://www.github.com/googleapis/python-network-security/commit/1cea5e5f2f171d57b9d08eb141270be9c3c9c805))


### Documentation


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==19.10b0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.9"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/network_security_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
